stdlib/LinearAlgebra/src/diagonal.jl

```diff
@@ -396,6 +396,86 @@ function _rdiv!(Dc::Diagonal, Db::Diagonal, Da::Diagonal)
 end
 ldiv!(Dc::Diagonal, Da::Diagonal, Db::Diagonal) = Diagonal(ldiv!(Dc.diag, Da, Db.diag))
 
+# optimizations for (Sym)Tridiagonal and Diagonal
+@propagate_inbounds _getudiag(T::Tridiagonal, i) = T.du[i]
+@propagate_inbounds _getudiag(S::SymTridiagonal, i) = S.ev[i]
+@propagate_inbounds _getdiag(T::Tridiagonal, i) = T.d[i]
+@propagate_inbounds _getdiag(S::SymTridiagonal, i) = symmetric(S.dv[i], :U)::symmetric_type(eltype(S.dv))
+@propagate_inbounds _getldiag(T::Tridiagonal, i) = T.dl[i]
+@propagate_inbounds _getldiag(S::SymTridiagonal, i) = transpose(S.ev[i])
+
+function (\)(D::Diagonal, S::SymTridiagonal)
+    T = promote_op(\, eltype(D), eltype(S))
+    du = similar(S.ev, T, length(S.dv)-1)
+    d  = similar(S.dv, T, length(S.dv))
+    dl = copy(du)
+    ldiv!(Tridiagonal(dl, d, du), D, S)
+end
+(\)(D::Diagonal, T::Tridiagonal) = ldiv!(similar(T, promote_op(\, eltype(D), eltype(T))), D, T)
+function ldiv!(T::Tridiagonal, D::Diagonal, S::Union{SymTridiagonal,Tridiagonal})
+    m = size(S, 1)
+    dd = D.diag
+    if (k = length(dd)) != m
+        throw(DimensionMismatch("diagonal matrix is $k by $k but right hand side has $m rows"))
+    end
+    if length(T.d) != m
+        throw(DimensionMismatch("target matrix size $(size(T)) does not match input matrix size $(size(S))"))
+    end
+    j = findfirst(iszero, dd)
+    isnothing(j) || throw(SingularException(j))
+    @inbounds begin
+        ddj = dd[1]
+        T.d[1] = ddj \ _getdiag(S, 1)
+        T.du[1] = ddj \ _getudiag(S, 1)
```

Comment:
We should skip L429~L438 if `m == 1`

Reply:
Certainly. That was probably giving silent BoundsErrors in the tests. I also handled the 0-length case now.

Reply:
At present, 0*0 sized `Tridiagonal` is illegal as we cant form a `-1` length AbstractVector. But `Tridiagonal` might support overlength like `SymTridiagonal` oneday, so I think we can let the 0-size check here.

Reply:
No, look at the tests. You can do
```julia
julia> Tridiagonal(Float64[], Float64[], Float64[])
0×0 Tridiagonal{Float64, Vector{Float64}}
```

Reply:
My fault, then we'd better fix `Diagonal(1:0) \ SymTridiagonal(1:0,1:0)` by replace `du = similar(S.ev, T, length(S.dv)-1)` with `du = similar(S.ev, T, max(0, length(S.dv)-1))`

Reply:
Good point!